### PR TITLE
Feature soundcloudcontentprovider

### PR DESCRIPTION
**Pull request by DavidChristiansen**
Hi guys,
Another content provider for your consideration. This time it's for soundcloud.com and the contentprovider spits out an HTML5 player.

As demonstrated here : http://jabbrnet.apphb.com/#/rooms/soundcloud

### CODE REVIEW

File: JabbR/ContentProviders/SoundCloudContentProvider.cs [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+			return new ContentProviderResultModel() 
+			{
+				Title = article.title,
+				Content = article.html
+			};
+		}
+
+		private static dynamic FetchArticle(Uri url)
+		{
+			var webRequest = (HttpWebRequest)WebRequest.Create(String.Format(_soundcloudAPIURL, url.AbsoluteUri));
```
Comment (sethwebster) [inline JabbR/ContentProviders/SoundCloudContentProvider.cs]: Is it possible to do this without the second request?

Comment (DavidChristiansen) [inline JabbR/ContentProviders/SoundCloudContentProvider.cs]: Let me look into this further.

Comment (DavidChristiansen) [inline JabbR/ContentProviders/SoundCloudContentProvider.cs]: Yes, it is

**Comment by davidfowl** [comment]
Looks like you're using tabs.

File: JabbR/ContentProviders/SoundCloudContentProvider.cs [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+		private static readonly Regex _titleRegex = new Regex("<meta.*content=\"(.*)\".*property=\"og:title\".*/>");
+		private static readonly Regex _trackIDExtractRegex = new Regex("<meta.*content=\"http://player\\.soundcloud\\.com/player\\.swf.*tracks%2F(.*?)&amp;.*property=\"og:video\"\\s*/>");
+
+		protected override ContentProviderResultModel GetCollapsibleContent(HttpWebResponse response)
+		{
+			using (var sr = new StreamReader(response.GetResponseStream())) 
+			{
+				var pageContent = sr.ReadToEnd();
+				if (string.IsNullOrEmpty(pageContent))
```
Comment (davidfowl) [inline JabbR/ContentProviders/SoundCloudContentProvider.cs]: No self closing ifs add braces.
Using String. not string.

File: JabbR/ContentProviders/SoundCloudContentProvider.cs [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+					.Groups
+					.Cast<Group>()
+					.Skip(1)
+					.Select(g => g.Value).SingleOrDefault(v => !String.IsNullOrEmpty(v));
+				
+				var titleContent = _titleRegex.Match(pageContent)
+					.Groups
+					.Cast<Group>()
+					.Skip(1)
+					.Select(g => g.Value).SingleOrDefault(v => !String.IsNullOrEmpty(v));
```
Comment (davidfowl) [inline JabbR/ContentProviders/SoundCloudContentProvider.cs]: Why do you have the same code 2 times in a row. Actually I find it strange that nobody has created a helper for this yet. How many times is this code repeated in content providers.

File: JabbR/ContentProviders/SoundCloudContentProvider.cs [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+using System.Text.RegularExpressions;
+using JabbR.ContentProviders.Core;
+using Newtonsoft.Json;
+
+namespace JabbR.ContentProviders
+{
+	public class SoundCloudContentProvider : CollapsibleContentProvider
+	{
+
+		private static readonly Regex _titleRegex = new Regex("<meta.*content=\"(.*)\".*property=\"og:title\".*/>");
```
Comment (davidfowl) [inline JabbR/ContentProviders/SoundCloudContentProvider.cs]: Don't like verbatim strings?

**Comment by DavidChristiansen** [comment]
Was written on my work PC so maybe the setting were asque

File: JabbR/ContentProviders/SoundCloudContentProvider.cs (hunk 4 above)
Comment (DavidChristiansen) [inline JabbR/ContentProviders/SoundCloudContentProvider.cs]: From the point of view of futureproofing, I don't want the regex to break should the spacing within this meta tag change. The regex is told to care about the three keypoints that are important - the tag meta, the attribute content and the property of og:title, everything else can change and the regex wont care.

**Comment by DavidChristiansen** [comment]
Please ignore last checkin - Something going wrong on the PC i was using and it is messing up either line ending or spacing.

**Comment by davidfowl** [comment]
Looks good in general but the braces are messed up.

**Pull request merged**